samples/nrf9160/modem_shell/src/location/location_service_nrf_cloud.c

```diff
@@ -0,0 +1,260 @@
+		err = -ENOMEM;
+		goto cleanup;
+	}
+	ret = cJSON_AddNumberToObject(data_obj, NRF_CLOUD_JSON_PGPS_GPS_DAY,
+				      pgps_req->gps_day);
+	if (ret == NULL) {
+		mosh_error("Failed to add GPS day to P-GPS request");
+		err = -ENOMEM;
+		goto cleanup;
+	}
+	ret = cJSON_AddNumberToObject(data_obj, NRF_CLOUD_JSON_PGPS_GPS_TIME,
+				      pgps_req->gps_time_of_day);
+	if (ret == NULL) {
+		mosh_error("Failed to add GPS time to P-GPS request");
+		err = -ENOMEM;
+		goto cleanup;
+	}
+
+	/* Convert CJSON object to string and send to nRF Cloud */
+	msg_string = cJSON_PrintUnformatted(pgps_req_obj);
+	if (msg_string == NULL) {
+		mosh_error("Could not allocate memory for request message");
+		goto cleanup;
+	}
+
+	struct nrf_cloud_tx_data mqtt_msg = {
+		.data.ptr = msg_string,
+		.data.len = strlen(msg_string),
+		.qos = MQTT_QOS_1_AT_LEAST_ONCE,
+		.topic_type = NRF_CLOUD_TOPIC_MESSAGE,
+	};
+
+	err = nrf_cloud_send(&mqtt_msg);
+	if (!err) {
+		mosh_print("P-GPS data requested");
+	}
+
+cleanup:
+	cJSON_Delete(pgps_req_obj);
+	cJSON_free(msg_string);
+
+	if (err) {
+		mosh_error("nRF Cloud P-GPS request failed, error: %d", err);
+	}
+}
+#endif /* CONFIG_NRF_CLOUD_PGPS */
+
+#if defined(CONFIG_LOCATION_METHOD_CELLULAR) || defined(CONFIG_LOCATION_METHOD_WIFI)
+static void location_service_nrf_cloud_location_ready_cb(
+	const struct nrf_cloud_location_result *const result)
+{
+	if ((result != NULL) && (result->err == NRF_CLOUD_ERROR_NONE)) {
+		nrf_cloud_location.latitude = result->lat;
+		nrf_cloud_location.longitude = result->lon;
+		nrf_cloud_location.accuracy = (double)result->unc;
+
+		k_sem_give(&location_ready);
+	} else {
+		if (result) {
+			mosh_error(
+				"Unable to determine location from cloud response, error: %d",
+				result->err);
+		}
+		/* Reset the semaphore to unblock location_service_nrf_cloud_pos_get()
+		 * and make it return an error.
+		 */
+		k_sem_reset(&location_ready);
+	}
+}
+
+static int location_service_nrf_cloud_pos_get(
+#if defined(CONFIG_LOCATION_METHOD_CELLULAR)
+	const struct lte_lc_cells_info *cell_data,
+#else
+	void *cell_data,
+#endif
+#if defined(CONFIG_LOCATION_METHOD_WIFI)
+	const struct wifi_scan_info *scanning_results,
+#else
+	void *scanning_results,
+#endif
+	struct location_data *const location)
+{
+	int err;
+	nrf_cloud_location_response_t callback = NULL;
+
+	k_sem_reset(&location_ready);
+
+	if (location) {
+		callback = location_service_nrf_cloud_location_ready_cb;
+	}
+
+	mosh_print("Sending positioning request to cloud via MQTT");
+	err = nrf_cloud_location_request(
+		cell_data, scanning_results, (callback != NULL), callback);
+	if (err == -EACCES) {
+		mosh_error("Cloud connection is not established");
+		return err;
+	} else if (err) {
+		mosh_error("Failed to request positioning data, error: %d", err);
+		return err;
+	}
+
+	if (callback) {
+		if (k_sem_take(&location_ready, K_MSEC(30000)) == -EAGAIN) {
+			mosh_error(
+				"Positioning data request timed out or "
```

Comment:
Maybe you could remove this error print and do all printing in the callback. This way you could be specific which was the cause for failure (no location or timeout).

Reply:
Timeout case won't trigger the callback, i.e., when the timer of this semaphore expires. And errors in cloud communication will trigger it and there is a print for those but I guess we want something for the timeout case too. It's not perfect.

Reply:
@hiltunent Close or comment further.

Reply:
Ok, I wasn't looking carefully enough.